ocaml/xapi/helpers.ml

```diff
@@ -1234,3 +1234,102 @@ let host_supports_hvm ~__context host =
 let env_with_path env_vars =
   Array.of_list @@
   List.map (fun (k,v) -> Printf.sprintf "%s=%s" k v) (("PATH", String.concat ":" Forkhelpers.default_path)::env_vars)
+
+module Task : sig
+  val wait_for : __context:Context.t -> tasks:API.ref_task list -> unit
+  val to_result : __context:Context.t -> of_rpc:(Rpc.t -> 'a) -> t:API.ref_task -> 'a
+end = struct (* can't place these functions in task helpers due to circular dependencies *)
+  let wait_for_ ~__context ~tasks ~propagate_cancel cb =
+    let our_task = Context.get_task_id __context in
+    let classes = List.map (fun x -> Printf.sprintf "task/%s" (Ref.string_of x)) (our_task::tasks) in
+    let rec process token =
+      (* if we have been cancelled, also cancel the tasks we are waiting on *)
+      if propagate_cancel && TaskHelper.is_cancelling ~__context then tasks |> List.iter (fun self -> TaskHelper.cancel_this ~__context ~self);
+      TaskHelper.exn_if_cancelling ~__context; (* First check if _we_ have been cancelled *)
+      let statuses = List.filter_map (fun task -> try Some (Db.Task.get_status ~__context ~self:task) with _ -> None) tasks in
+      let unfinished = List.exists (fun state -> state = `pending) statuses in
+      if unfinished
+      then begin
+        let from = call_api_functions ~__context
+            (fun rpc session_id -> Client.Client.Event.from ~rpc ~session_id ~classes ~token ~timeout:30.0) in
+        debug "Using events to wait for tasks: %s" (String.concat "," classes);
+        let from = Event_types.event_from_of_rpc from in
+        cb (); (** be careful not to include logic in `cb` that could modify tasks we are waiting on indefinitely *)
```

Comment:
realised that we have to be careful here to not unnecessarily change the tasks we are listening for events on

as a precaution, i made `mirror_progress` more future proof - we don't want to mirror our progress on our own task